Adding `Target` and `InstructionProperties` representations to rust

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #12052

~~**Currently blocked by #12459**~~ thanks @mtreinish!

As we continue moving more parts of the transpiler to Rust, it is important that some of the important data about the devices we're compiling for is easily accessible from the Rust side of things. These commits aim to implement representations of  both the `Target` and `InstructionProperties` classes, allowing most of their data to live inside of Rust but still be usable in Python.

### Details and comments

`qiskit._accelerate` now has a `target` module which includes both the `BaseTarget` and `BaseInstructionProperties` structs.
Here's a more detailed description of what has changed:

#### Rust

- `BaseTarget` will include the main data that the `Target` has in python. This will include all of the publicly available attributes of the existing Target class, except for the following:
   - `_coupling_graph`: of type `rustworkx.PyDiGraph`
   - `_instruction_durations`: of type `InstructionDurations`
   - `_instruction_schedule_map`: of type `InstructionScheduleMap`

   Due to these properties are not usable through Rust without using python api's, they were left in python.

   - This representation is purely additive through Python and non mutable through Rust.

- `BaseInstructionProperties` represents the main data of `InstructionProperties, that being:
   - Error: `f64`,
   - Duratlon: `f64`

#### Python:
- `Target`: this is a subclass of `BaseTarget` and has the missing attributes + functions from the previous version.
   - In addition it stores a copy of the `_gate_map` attribute from the rust side for easy access in python without conversions from rust.
   - We ensure that the data from `Target._gate_map` is identical to the data in `super().gate_map` by calling the `super()`'s equivalent function when adding or updating an instruction.
- `InstructionProperties`: is a subclass of `BaseInstructionProperties`
    - The attribute of `calibration` to store the instruction's calibration object is present in this representation along with it's custom setter and getter.

#### Possible breaking changes:
~~- `InstructionProperties` is now PyO3 class which means it cannot be subclassed by overloading `__init__()` alone. The new architecture would require users to extend both `__new__` and `__init__` to add extra properties. These wouldn't necessarily be visible in rust.~~ **Fixed!**

#### Broken tests
~~A procedure performed in Aer that attempts to erase the `Target`'s internal data is no longer possible due to the these attributes being hidden from python.~~ Solved by Qiskit/qiskit-aer#2142

#### Blocks
- [x] raynelfss#2
- [x] #12697
- [x]  #12698

#### Open to review!
Any suggestions, comments, feedback, or any request for changes are encouraged and welcome!

raynelfss committed:
Initial: Add `Target` class to `_accelerate`
- Add `Target` class to test mobility between Rust and Python.
- Add `add_instruction` method to test compatibility with instructions.

raynelfss committed:
Fix: Remove empty property check
- Property check caused most cases to panic.
- Will be commented out and restored at a later time.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Add: Instructions property
- Instructions property returns all added to the target.
- Similar behavior to source.

raynelfss committed:
Chore: comments and deprecated methods
- Add comments to instruction property.
- Use new_bound for new PyDicts.

raynelfss committed:
Chore: Remove redundant code
- Remove redundant transformation of PyObject to PyTuple.
- Remove debugging print statement.

raynelfss committed:
Add: `InstructionProperties` class and type checkers
- Add `InstructionProperties` class to process properties in rust.
- Add `is_instance` and `is_class` to identify certain Python objects.
- Modify logic of `add_instruction` to use class check.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Add: Setter and Getter for calibration in `InstructionProperty`

raynelfss committed:
Add: `update_instruction_properties` to Target.

raynelfss committed:
Add: Update_from_instruction_schedule_map
- Partial addition from Target.py\
- Introduction of hashable qarg data structure.
- Other tweaks and fixes.

raynelfss committed:
Add: Complete `update_from_instruction_schedule_map1
- Complete missing procedures in function.
- Rename `Qargs` to `HashableVec`.
- Make `HashableVec` generic.
- Separate `import_from_module_call` into call0 and call1.
- Other tweaks and fixes.

raynelfss committed:
Add: instruction_schedule_map property.
- Remove stray print statements.
- Other tweaks and fixes.

raynelfss committed:
Fix: Key issue in `update_from_instruction_schedule_map`
- Remove all unsafe unwraps

raynelfss committed:
Fix: Use PyResult Value for void functon
- Update `update_from_instruction_schedule_map to use PyResult and '?' operator.
- Use Bound Python objects whenever possible.
- Other tweaks and fixes.

raynelfss committed:
Add: Python wrapping for Target
- Add temporary _target module for testing.
- Remove update_from_instruction_schedule_map function back to python.
- Add python properties for all public attributes in rust
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Add: `qargs` property
- Add identical method `qargs` to obtain the qargs of a target.
- Other tweaks and fixes.

raynelfss committed:
Add: `qargs_for_operation_name` function.
- Add function with identical behavior to the original in Target.
- Other tweaks and fixes.

raynelfss committed:
Add: durations method for Target
- Add target module to qiskit init file.
- Remove is_instance method.
- Modify set_calibration method in InstructionProperty to leave typechecking to Python.
- Change rust Target alias to Target2.
- Other tweaks and fixes,

raynelfss committed:
Add: InstructionProperties wrapper in python

raynelfss committed:
Fix: InstructionProperties could not receive calibrations
- Fix wrong setters/getters for calibration in InstructionProperty object in rust.

raynelfss committed:
Add: more methods to Target in `target.rs`
- Add FromPyObject trait to Hashable vec to receive Tuples and transform them directly into this type.
- Add operations_for_qargs for Target class in Rust side and Python.
- Fix return dict keys for `qargs_for_operation_name`.
- Add `timing_constrains` and `operation_from_name` to Python side.
- Other tweaks and fixes.

raynelfss committed:
Fix: missing return value in `operations_for_args`
- Fix wrong name for function operation_for_qargs.
- Fix missing return value in the python side.
- Other tweaks and fixes.

raynelfss committed:
Fix: Bad compatibility with InstructionProperties
- Make `InstructionProperties` "_calibration" attribute visible.
- Removed attribute "calibration", treat as class property.
- Other tweaks and fixes

raynelfss committed:
Add: `operation_names_for_qargs` to Target
- Port class method to rust and connect to Python wrapper.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Add: instruction_supported method to rust and python:
- Other tweaks and fixes.

raynelfss committed:
Add: changes to add_instruction function to increase functionality.
- These changes break current functionality of other functions, butemulate intended behavior better.
- Fixes coming soon.

raynelfss committed:
Fix: Backwards compatibility with `add_instruction`
- Fixed wrong additions to HashMaps in the rust side causing instructions to be missing.
- Other tweaks and fixes.

raynelfss committed:
Fix: Gate Map behavior didn't match #11422
- Make GateMap use optional values to match behavior of #11422.
- Define GateMapType for complex type in self.gate_map.
- Throw Python KeyError exceptions from the rust side in `update_instruction_properties` and other functions.
- Modify logic in subsequent functions that use gate_map optional values.
- Other tweaks and fixes.

raynelfss committed:
Add: `has_calibration` method to Target

raynelfss committed:
Add: `get_calibraton` method to Target

raynelfss committed:
Add: `instruction_properties` method to Target

raynelfss committed:
Add: `build_coupling_map` and helper methods
- `build_coupling_map`will remain in Python for now, along with its helper functions.
- Make `gate_name_map` visible to python.
- Add `coupling_graph` attribute to Target in Rust.
- Other tweaks and fixes.

raynelfss committed:
Add: `get_non_global_operation_names` to Target.
- Add attributes `non_global_strict_basis` and `non_global_basis` as Optional.
- Other tweaks and fixes.

raynelfss committed:
Add: Missing properties
- Add properties: operations, operation_names, and physical_qubits.
- Reorganize properties placement.
- Other tweaks and fixes.

raynelfss committed:
Add: `from_configuration` classmethod to Target.
- Add method that mimics the behavior of the python method.
- Change concurrent_measurements to 2d Vec instead of a Vec of sets.
- Other tweaks and fixes.

raynelfss committed:
Add: Magic methods to Rust and Python
- Add docstring to __init__.
- Add __iter__, __getitem__, __len__, __contains__, keys, values, and items methods to rust.
- Add equivalen methods to python + the __str__ method.
- Make description an optional attribute in rust.
- Other tweaks and fixes.

raynelfss committed:
Fix: Bugs when fetching qargs or operations
- Fix qarg_for_operation_name logic to account for None and throw correct exceptions.
- Stringify description before sending in case of numerical descriptors.
- Fix qarg to account for None entry.
- Other tweaks and fixes.

raynelfss committed:
Chore: Prepare for Draft PR
- Remove _target.py testing file.
- Fix incorrect initialization of calibration in InstructionProperties.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: target not being recognized as a module
- Add target to the pyext crate.
- Change placement of target import for alphabetical ordering.
- Other tweaks and fixes.

raynelfss committed:
Fix: Change HashMap to IndexMap
- Change from f32 to f64 precision.
- Other tweaks and fixes.

raynelfss committed:
Fix: Move InstructionProperties fully to Rust
- Move InstructionProperties to rust.
- Modify gate_map to accept an InstructionProprties object instead of PyObjecy.
- Change update_instruction_properties to use Option InstructionProprtyird.
- Remove InstructionProperties from target.py
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Fix: Make Target inherit from Rust
- Make Target inherit from the rust-side Target by using subclass attribute, then extending its functionality using python.
- Switch from __init__ to __new__ to adapt to the Target rust class.
- Modify every trait that worked with `target._Target` to use `super()` or `self` instead.
- Fix repr in InstructionProperties to not show `Some()` when a value exists.
- Fix `__str__` method in `Target` to not display "None" if no description is given.
- Assume `num_qubits` is the first argument when an integer is provided as a first argument and nothing else is provided for second (Target initializer).
- Return a set in operation_names instead of a Vec.
- Other tweaks and fixes.

raynelfss committed:
Fix: Recognize None in `operation_for_qargs`.
- Fix module labels for each class in target.rs.
- Use py.is_instance instead of passing isinstance to `instruction_supported`.
- Modify `operations_for_qargs` to accept optional values less aggressively. Allow it to find instructions with no qargs. (NoneType).
- Other tweaks and fixes.

raynelfss committed:
Fix: Make InstructionProperties subclassable.
- Fix get_non_global_operation_names to accept optional values and fix search set to use sorted values.
- Fix __repr__ method in InstructionProperties to add punctuation.
- Fix typo in python durations method.
- Modify test to overload __new__ method instead of just  __init__ (Possible breaking change).
-Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Fix: errors in `instruction_properties` and others:
- Allow `instruction_properties` method to view optional properties.
- Allow `operation_names_for_qargs` to select class instructions when None is passed as a qarg.
- Modify __str__ method to display error and duration times as int if the value is 0.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Fix: call `isclass` from rust, instead of passing it from Python.

raynelfss committed:
Fix: Move `update_from_instruction_schedule_map` to rust.

raynelfss committed:
Fix: Move `durations` to rust.

raynelfss committed:
Fix: Move `timing_constraints` to rust

raynelfss committed:
Fix: Move operations_from_name fully to rust

raynelfss committed:
Fix: `instruction_supported` method:
- Rewrite the logic of instruction_supported due to previous errors in the method.
- Move `check_obj_params` to Rust.
- Other tweaks and fixes.

raynelfss committed:
Fix: errors in `from_configuration` class method.
- Fix some of the logic when retrieving gates from `name_mapping`.
- Remove function arguments in favor of implementing counterpart functions in rust.
- Add qubit_props_list_from_props function and return rust datatypes.
- Fix wrong error handling procedures when retrieving attributes from backend_property.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Fix: Import `InstructionScheduleMap` directly instead of passing.
- `instruction_schedule_map()` now imports the classtype directly from rust instead of needing it to be passed from python.
- Remove unused imports in `target.py`.
- Ignore unused arguments in `test_extra_props_str`.
- Other tweaks and fixes.

raynelfss committed:
Docs: Add docstrings to rust functions
- Remove redundant redefinitions in python.
- Fix text_signatures for some rust functions.
- Added lint exceptions to some necessary imports and function arguments.
- Other tweaks and fixes.

raynelfss committed:
Add: Make `Target` and `InstructionProperties` pickleable.
- Add `__getstate__` and `__setstate__` methods to make both rust subclasses pickleable.

raynelfss committed:
Fix: Wrong calibration assignment in __setstate__
- Use set_calibration to set the correct calibration argument.
- Fix wrong signature in get_non_global_operation_names.
- Other tweaks and fixes.

raynelfss committed:
Refactor: HashableVec is now Qarg
- Use `PhysicalQubit` instead of u32 for qargs.
- Use a `SmallVec` of size 4 instead of a dynamic Vec.
- Default to using the `Hash()` method embedded in `SmallVec`.
- Add a Default method to easily unwrap Qarg objects.
- Other tweaks and fixes.

raynelfss committed:
Add: `get` function to target.
- Remove some redundant cloning in code.
- Other small fixes.

raynelfss committed:
Fix: Remove unnecessary Optional values in gate_map.
- Update every gate_map call to use the new format.
- Other small tweaks and fixes.

raynelfss committed:
Refactor: `calibration` is for `InstructionProperties`
- Use python `None` instead of option to store `calibration` in `InstructionProperties`.
- Adapt code to these changes.
- Remove redundant implementation of Hash in Qargs.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Temporary speedup for `gate_map` access
- Added temporary speedups to access the gate_map by returning the values as PyObjects.
- Convert qargs to rust tuples instead of initializing a `PyTuple`.
- Store `InstructionProperties` as a python ref in gate_map. (Will be changed in future updates).
- Other tweaks anf fixes.

raynelfss committed:
Fix: Incorrect extractions for `InstructionProperties`
- Fix incorrect conversion of `InstructionProperties` to `Py<InstructionProperties>`
- Fix incorrect extraction of qargs in `update_from_instruction_schedule_map`

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Hide all private attributes in `Target`
- Hide all private attributes of the `Target` to prevent unecessary cloning.
- Other small tweaks and fixes.

raynelfss committed:
Add: New representation of gate_map using new pyclasses:
- Make Qarg a sequence pyclass.
- Make QargPropsMap the new representation of a GateMap value.
- Adapt the code to new structure.
- TODO: Add missing magic methods for sequence and mapping objects.
- Other small tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Add: Use custom datatypes to return values to Python.
- Add QargSet datatype to return a set of qargs.
   - Works as return type for `Target.qargs`
   - Object is has itertype of QargSetIter.
- Rename QargPropMap to PropsMap
   - Use iterator type IterPropsMap
- Other small tweaks and fixes.

raynelfss committed:
Fix: Extend `InstructionProperties` to be subclassable using `__init__:
- Made a subclass of `InstructionProperties` that can be extended using an `__init__`method.
- Revert previous changes to `test_target.py`.
- Other tweaks and fixes.

raynelfss committed:
Refactor: Split target into its own module
- Reestructure the files to improve readability of code.
   - `instruction_properties.rs` contaisn the `InstructionProperties` class.
   - `mod.rs` contains the `Target` class.
   - `qargs.rs` contains the Qargs struct to store quantum arguments.
   - `property_map` contains the Qarg: Property Mapping that will be stored in the gate_map.
- Add missing methods to PropsMap:
   - Add `PropsMapKeys` object to store the qargs as a set.
   - Add methods to compare and access `PropsMapKey`.
- Add QargsOrTuple enum in Qargs to parse Qargs instantly.

raynelfss committed:
Fix: Rest of failing tests in Target
- Modify the `InstructionProperties` python wrapper.
   - InstructionProperties was not assigning properties to rust side.
- Make duration in `InstructionProperties` setable.
- Add `__eq__` method for `PropMap` to compare with other dicts.
- `PropMapKeys` can only be compared with a Set.
- Remove `qargs_for_operation_name` from `target.py`
- Other small tweaks and fixes.

raynelfss committed:
Add: New GateMap Structure
- GateMap is now its own mapping object.
- Add `__setstate__` and `__getstate__` methods for `PropMap` and `GateMap`.
- Other small tweaks and fixes.

raynelfss committed:
Fix: Make new subclasses pickleable
- Add module location to `PropsMap`, `GateMap`, and `Qargs`.
- Added default method to PropMap.
- Made default method part of class initializers.
- Other smalls tweaks and fixes.

raynelfss committed:
Fix: Remove redundant lookup in Target (#12373)

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Format: `mod.rs` qubit_comparison to one line.

raynelfss committed:
Add: `GateMapKeys` object in GateMap:
- Use IndexSet as a base to preserve the insertion order.
- Other tweaks and fixes.

raynelfss committed:
Add: __sub__ method to GateMapKeys

raynelfss committed:
Fix: Modify `GateMap` to store values in Python heap.
- Fix `GateMap.__iter__` to use an IndexKeys iterator.
- Other small tweaks and fixes.

raynelfss committed:
Fix: Remove duplicate import of `IndexSet::into_iter` in `GateMap`.
- Make `__iter__` use the keys() method in `GateMap`.

raynelfss committed:
Fix:: Adapt to target changes (#12288)
- Fix lint stray imports.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Incorrect creation of parameters in `update_from_instruction_schedule_map`
- Add `tupelize` function to create tuples from non-downcastable items.
- Fix creation of Parameters by iterating through members of tuple object and mapping them to parameters in `update_from_instruction_schedule_map`.
- Add missing logic for creating a Target with/without `qubit_properties`.
- Add tuple conversion of `Qargs` to store items in a dict in `BasisTranslator` and `UnitarySynthesis` passes.
- Cast `PropsMap` object to dict when comparing in `test_fake_backends.py`.
- Modify logic of helper functions that receive a bound object reference, a second `py` not required as an argument.
- Add set operation methods to `GateMapKeys`.
- Other tweaks and fixes.

raynelfss committed:
Fix: More failing tests
- Fix repeated erroneous calls to `add_instruction` in `update_from_instruction_schedule_map`
- Add missing condition in `instruction_supported`
- Use `IndexSet` instead of `HashSet` for `QargsSet`.
- Other small tweaks and fixes.

raynelfss committed:
Add: Macro rules for qargs and other sequences.
- Create `QargSet` and `PropsMap` using the new macros.
- Return a `TargetOpNames` ordered set to python in `operation_names`.
- Remove the Python side `operation_names.`
- Fix faulty docstring in `target.py`.
- Other tweaks and fixes.

raynelfss committed:
Docs: Add necessary docstrings to all new rust functions.
- Remove duplicate Iterator in GateMap.
- Other small tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Use `GILOneCell` and remove `Qargs`
- Use `GILOneCell` to import python modules only once at initialization.
- Remove the custom data structure `Qargs` to avoid conversion overhead.
- `Qargs` does not use `PhysicalQubits`, `u32` is used instead.
- Fix `__setstate__ `and `__getstate__` methods for `PropsMap`, `GateMap`, and `key_like_set_iterator` macro_rule.
- Update code to use the new structures.
- TODO: Fix broken tests.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Cast `Qargs` to `Tuple` in specific situations
- Use tupleize to cast `Qargs` to `Tuple` in `instructions`.
- Use downcast to extract string in `add_instruction`.
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss commented:
I wonder if I should wait for #11461  for this to continue, since it introduces a change to the way ``Target.from_configuration()`` works.

raynelfss committed:
Add: Make `Target` Representable in Rust
- Rename `InstructionProperties` as `BaseInstructionProperties`.
   - Remove `Calibration` from the rust space.
- Restore `gate_map`, `coupling_map`, `instruction_schedule_map`, and `instruction_durations` to rust.
- Remove all unnecessary data structures from rust space.
- Other tweaks and fixes.

raynelfss committed:
Refactor: Remove previour changes to unrelated files.

raynelfss committed:
Add: rust native functions to target
- Added rust native functionality to target such that a `py` would not be needed to use one.
- Add Index trait to make `Target` subscriptable.
- Other small tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Remove all unnecessary python method calls.
- Remove uage of `inspect.isclass`.
- Rename `Target` to `BaseTarget` in the rust side.
- Rename `err.rs` to `errors.rs`.
- Remove rust-native `add_inst` and `update_inst` as Target should not be modified from Rust.
- Made `add_instruction` and `update_instruction_properties` private in `BaseTarget`.
- Add missing `get` method in `Target`.
- Other tweaks and fixes

raynelfss committed:
Format: Fix lint

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Wrong return value for `BaseTarget.qargs`

mtreinish commented on crates/accelerate/src/target_transpiler/instruction_properties.rs:
```diff
@@ -0,0 +1,78 @@
+*/
+#[pyclass(subclass, module = "qiskit._accelerate.target")]
+#[derive(Clone, Debug)]
+pub struct BaseInstructionProperties {
```
A suggestion on naming, from Rust space I think we should call this `InstructionProperties` because it's a rust representation of the instruction properties, there isn't anything building off of it in rust space so `Base` is more confusing. We can set a python space name in the `pyclass` macro to make it `BaseInstructionProperties` for python where the name makes more sense.

```suggestion
#[pyclass(subclass, name = "BaseInstructionProperties", module = "qiskit._accelerate.target")]
#[derive(Clone, Debug)]
pub struct InstructionProperties {
```

raynelfss committed:
Add: Temporary Instruction representation in rust.
- Add temporary instruction representation to avoid repeated extraction from python.

raynelfss committed:
Add: Native representation of coupling graph

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Wrong attribute extraction for `GateRep`

raynelfss committed:
Remove: `CouplingGraph` rust native representation.
- Move to different PR.

raynelfss committed:
Format: Remove stray whitespace

raynelfss committed:
Add: `get_non_global_op_names` as a rust native function

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+}
+
+// Custom types
+type Qargs = SmallVec<[PhysicalQubit; 4]>;
```
Why `4`? In general most targets will have 1 and 2 qubit qargs. So I would set this to `2` to optimize for the common case.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+// Custom types
+type Qargs = SmallVec<[PhysicalQubit; 4]>;
+type GateMap = IndexMap<String, PropsMap>;
+type PropsMap = IndexMap<Option<Qargs>, Option<BaseInstructionProperties>>;
```
Should you set the hasher here to be ahash?

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+    pub num_qubits: Option<usize>,
+    pub label: Option<String>,
+    pub params: Option<Vec<Param>>,
+}
```
Is this temporary in that you're waiting for #12459 to merge to finalize it? What were you thinking this will look like in it's final form?

raynelfss replied:
Also temporary, this makes it easier to access properties of the gate without having to extract them every time. But it will adapt better to whatever changes your PR includes.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+pub enum Param {
+    Float(f64),
+    ParameterExpression(PyObject),
+}
```
I assume this is based on what I had in #12459, if so I did have to change definition here to be a 3 entry `enum` that also accounted for other types of parameters (since non-standard gates can contain anything as a parameter). I also needed to custom implement `FromPyObject` for this.

raynelfss replied:
Alright, I'll just wait for the one used in your PR. Since it requires certain imports that are only present there.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+    an existing object and create a new subset (or use one of the methods
+    to do this). The object internally caches different views and these
+    would potentially be invalidated by removals.
+ */
```
Do we need this docstring still? I'm wondering if a rust native docstring would make more sense, since while this is the base class for `Target` it's not actually something user facing in python. The advantage of having this base class live in rust is that we can get at the data from Rust now. So maybe documenting how this works in the rust domain will make it easier for us to use in Rust.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+    _gate_name_map: IndexMap<String, GateRep>,
+    global_operations: IndexMap<usize, HashSet<String>>,
+    variable_class_operations: IndexSet<String>,
+    qarg_gate_map: IndexMap<Option<Qargs>, Option<HashSet<String>>>,
```
I'd probably use ahash for the hasher on all of these, just to speed up the lookups a bit.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+// Rust native methods
+impl BaseTarget {
+    /// Generate non global operations if missing
+    fn generate_non_global_op_names(&mut self, strict_direction: bool) -> &Vec<String> {
```
It's typical more idomatic to return a slice like `&[String]` instead of a borrowed Vec. Is there are reason you needed to return the full Vec?

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+    #[pyo3(get, set)]
+    pub acquire_alignment: i32,
+    #[pyo3(get, set)]
+    pub qubit_properties: Option<Vec<PyObject>>,
```
You left this as a `PyObject` to avoid having to define a struct for this in Rust? I think this is fine for a temporary thing but we'll want to represent this in Rust eventually (it's not critical for this PR). Can you add a `TODO` comment here to track a follow on work item to move this to rust.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+                    .or_insert(Some(HashSet::from([name.clone()])));
+            }
+        }
+        // Modify logic once gates are in rust.
```
Just so we have a common keyword to search for.
```suggestion
        // TODO: Modify logic once gates are in rust.
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+}
+
+// To access the Target's gate map by gate name.
+impl Index<&String> for BaseTarget {
```
Can you use `&str` here to make this a bit more generic? Or does it have to be `&String`?

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+            for qarg_key in self.qarg_gate_map.keys().flatten() {
+                if qarg_key.len() != 1 {
+                    let mut vec = qarg_key.clone();
+                    vec.sort();
```
You can probably use `sort_unstable()` here since there will never be duplicates.

```suggestion
                    vec.sort_unstable();
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+impl BaseTarget {
+    /// Generate non global operations if missing
+    fn generate_non_global_op_names(&mut self, strict_direction: bool) -> &Vec<String> {
+        let mut search_set: HashSet<Option<Qargs>> = HashSet::new();
```
With `None` there can't be any other constraints on it because that means it's globally defined. This probably can be simplified if you changed search set to just be HashSet<Qargs> and you detect and  `None` and skip it.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1238 @@
+    }
+
+    fn __getstate__(&self, py: Python<'_>) -> PyResult<Py<PyList>> {
+        let result_list = PyList::empty_bound(py);
```
Since there are so many fields here I'd probably use a dict, so that you can access by name. With 15 elements tracking which index maps to which field seems quite error prone when we need to update this.

mtreinish reviewed (commented):
Thanks for the hard work on this, I've done a quick first pass through most of the code, still have a bit more to get through, but figured I'd leave me comments so far so you can see my thoughts at this point.

I also have a couple of high level questions I have is how were you planning to work with this in conjunction with #12459? Were you think go forward with the temporary structures you've defined and integrate after the fact, or wait until #12459 merges and rebase on top of that? I feel like whatever we end up with in #12459 will inform your design here.

The other musing I have in the back of my head is thinking about whether this hybrid Python and Rust structure is the approach we want to go with. I've been thinking for some time it might make sense to just make a Rust native Target struct that contains the information we need from a target natively in Rust.  Unlike for gates/circuit and the dag the target is unique that when we're working with it in Rust we never need to mutate it. So doing this split approach where we have a read-only rust space view of the data from a Python target would be fine. We could build this struct from Python on demand and cache it in the python object so that we only build it once (and invalidate  the cached item on subsequent target modification in Python) The reason I'm thinking this is while this hybrid approach works, I worry we're tying a few too many Python-isms into this struct. For example, `instruction_supported()` is returning a `PyResult` and from a rust perspective I feel like this method should be infallible, a target either supports a given instruction or not. What we really need here is to have a convenient and ergonomic way for transpiler passes to query the constraints from a target directly in rust without having to worry about Python.  That being said if we went with what I'm suggesting we'd end up with duplicate implementations so it's not a 100% clear cut decision for me.

raynelfss committed:
Fix: Use Ahash for Hashing
- Use ahash for hashing when possible.
- Rename `BaseTarget` to `Target` in rust only.
- Rename `BaseInstructionProperties` to `InstructionProperties` in rust only.
- Remove optional logic from `generate_non_global_op_names`.
- Use dict for `__setstate__` and `__getstate__` in `Target`.
- Reduced the docstring for `Target` and `InstructionProperties`.
- Other small tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Format: new changes to `lib.rs`

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Format: Adapt to new lint rules

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Use new gates infrastructure (#12459)
- Create custom enum to collect either a `NormalOperation` or a `VariableOperation` depending on what is needed.
- Add a rust native `is_instruction_supported` method to check whether a Target supports a certain instruction.
- Make conversion methods from `circuit_instruction.rs` public.
- Add comparison methods for `Param` in `operations.rs`
- Remove need for isclass method in rustwise `add_instruction`
- Other tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Format: Fix rust formatting

raynelfss committed:
Add: rust-native method to obtain Operstion objects.

raynelfss committed:
Add: Comparison methods for `Param`

raynelfss committed:
FIx: Add display methods for `Params`

raynelfss committed:
Format: Fix lint test

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Format: Wrong merge conflict solve

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Improve rust methods to use iterators.
- Adapt the Python methods to leverage the rust native improvements.
- Use python native structures for the Python methods.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Format: Remove extra blankspace

raynelfss committed:
Fix: Remove `text_signature`, use `signature` instead.

raynelfss committed:
Fix: Rectify the behavior of `qargs`
- Keep insertion order by inserting all qargs into a `PySet`.
- Perform conversion to `PyTuple` at insertion time leveraging the iterator architecture.
- Remove python side counterpart to avoid double iteration.
- Make rust-native `qargs` return an iterator.

raynelfss committed:
Merge branch 'main' into move-target

mtreinish commented on crates/circuit/src/operations.rs:
```diff
@@ -184,6 +184,43 @@ impl ToPyObject for Param {
+            _ => false,
+        }
+    }
+}
```
I'd be worried about the overhead here. If we need to do this across thread boundaries the `with_gil()` in `compare()` will serialize execution and be generally really slow. We can probably keep `compare()` and give it a `py: Python` argument so we can control how the gil is handled. But we should be making it explicit here.  I'm a bit worried about this with the `Operation` trait too, and was planning to investigate potentially rework it once we start having parallelism in passes. But for something like `PartialEq` which is in the stdlib I feel the risk is a lot higher because we might not even realize this.

raynelfss replied:
This will be removed based on Jake's comments of #12697, I am making updates to this.

mtreinish commented on crates/circuit/src/operations.rs:
```diff
@@ -184,6 +184,43 @@ impl ToPyObject for Param {
+        .unwrap_or("None".to_owned());
+        write!(f, "{}", display_name)
+    }
+}
```
Is this needed for anything? If not lets save it for a standalone PR and discuss it in that.

raynelfss replied:
This will also be removed as per #12697

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -220,38 +238,19 @@ class Target(Mapping):
-        "qubit_properties",
-        "_global_operations",
-        "concurrent_measurements",
-    )
```
Are slots incompatible with being based on a rust pyclass? If not I would suggest leaving the slots for the python space attributes defined in this class so that the python space component of the class is slotted still.

raynelfss replied:
I will add slots for those attributes that live in Python space, it seems that extending a `PyClass` does not accept slots for that class.

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -295,37 +294,45 @@ def __init__(
+        if description is not None:
+            if num_qubits is None and isinstance(description, int):
+                num_qubits = description
+                description = None
```
I see the intent here and it's to correct a lot of longstanding behavior around the API implications of this. While I get that a lot of the tests are doing `Target(2)` or something similar (especially since I wrote those) this is changing behavior. Like what would happen if a user did `Target(2, num_qubits=4)` then what would you do? I'd say hold off on doing this just for the sake of compatibility. If we want to make this change we should do it in isolation so we can discuss the backwards compatibility implications of doing this.

raynelfss replied:
Got it, it also reminds me of what @1ucian0 said about changing the behavior based on typing. This will be removed in future commits.

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -525,7 +509,7 @@ def update_from_instruction_schedule_map(self, inst_map, inst_name_map=None, err
                 try:
                     # Update gate error if provided.
-                    props.error = error_dict[inst_name][qargs]
+                    setattr(props, "error", error_dict[inst_name][qargs])
```
Is there a reason to change this syntax? Using `setattr()` is fine, but I don't see any difference with this since error always has to exist right?

raynelfss replied:
Yes, there should be no difference, this was old code that behaved based on having `InstructionProperties` be wrapped in a `Py` reference. This will be promptly moved back to its original syntax.

Edit: Okay, so the actual reason this was done was more because of formatting, PyLint was complaining about invisible attributes since Python does not know of the existence of the rust-only attributes and I had to make it this way.

mtreinish commented on crates/accelerate/src/target_transpiler/instruction_properties.rs:
```diff
@@ -0,0 +1,76 @@
+            output.push_str(", ");
+        } else {
+            output.push_str("error=None, ");
+        }
```
While not performance critical I would try to do this as a single `format!()` call because each `push_str()` will potentially result in another allocation to expand the string. You can just move the if statements to the top of the method and generate the str to substitute into `format!()`.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+#[derive(Debug, Clone, FromPyObject)]
+pub enum TargetOperation {
+    Normal(NormalOperation),
+    Variable(VariableOperation),
```
For `Variable` here are you referring to operations which can take a variable number of qubits? If so the normal term we use for that is "variadic" (https://en.wikipedia.org/wiki/Variadic_function)

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    fn num_qubits(&self) -> u32 {
+        match &self {
+            Self::Normal(normal) => normal.operation.num_qubits(),
+            Self::Variable(_) => 0,
```
Is this correct? Or are you using 0 as a sentinel value here? Maybe I don't understand what `Variable` is here.

raynelfss replied:
So `Variable` is when we provide a class instead of a class instance because it could have a variable amount of qubits (I believe we only do this for `ControlFlowOps`). And so if you call for `num_qubits` it will return a `@property`. We typically wouldn't have a need to call for this on a `VariableOp` as the `Target` operates on them differently, so I could just make this `unreachable!()`.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+pub struct NormalOperation {
+    operation: OperationType,
+    params: SmallVec<[Param; 3]>,
+    internal_name: String,
```
Is `internal_name` the name as defined in the target? I was a little confused by "internal" here, like I wasn't sure if you meant the name from the `OperationType` object or internal to the target.

raynelfss replied:
This will be removed, it was a way of comparing instructions by class names instead of by instance, in the case we had to.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    #[pyo3(get, set)]
+    pub pulse_alignment: i32,
+    #[pyo3(get, set)]
+    pub acquire_alignment: i32,
```
Is there a reason you used `i32` here? These fields can't really be negative as they're just an integer from from 1 to n which are a multiple of `dt`. I would use `u32` for all of these, but we can probably get away with `u8` or `u16` realistically since I don't think this will typically be > 255 or >65k. But maybe @nkanazawa1989 has some thoughts on this.

raynelfss replied:
I'll set it to `u32` fore now, but will wait on further conversation on this.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    // TODO: Port to Rust.
+    pub qubit_properties: Option<Vec<PyObject>>,
+    #[pyo3(get, set)]
+    pub concurrent_measurements: Vec<Vec<usize>>,
```
This might make more sense, since it's a grouping of qubit indices.
```suggestion
    pub concurrent_measurements: Vec<Vec<PhysicalQubit>>,
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+            pulse_alignment: pulse_alignment.unwrap_or(1),
+            acquire_alignment: acquire_alignment.unwrap_or(0),
+            qubit_properties,
+            concurrent_measurements: concurrent_measurements.unwrap_or(Vec::new()),
```
Was there any checking or semantics around the usage of this with `None`? Or is an empty list treated the same way in the python side?

raynelfss replied:
Seems to work the same with None as it does with an empty vec. I will change it to None as it was in the original `Target`.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+                }
+            }
+        }
+        // TODO: Modify logic once gates are in rust.
```
Gates are in rust now, is this still relevant?

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    ///     AttributeError: If gate is already in map
+    ///     TranspilerError: If an operation class is passed in for ``instruction`` and no name
+    ///         is specified or ``properties`` is set.
+    #[pyo3(signature = (instruction, name, properties=None))]
+    fn add_instruction(
```
I'm wondering if we want to make this different somehow. I was initially confused because the signature here is different from the python space methods. and I thought at first this was being exposed to end users (which would have been a backwards incompatible change). This might just be an issue with my review since I'm trying to keep too much state in my head at once.

raynelfss replied:
It's a little different from the python one because this function is overloaded in the python side. Whereas in python providing a name could be skipped, here it won't be needed due to Python already taking care of that already. Since we spoke of just making the Target only modifiable through Python and just having its data be accessible in rust. The `add_instruction` and `update_instruction_properties` methods are only exposed to python and not rust. And there are extra checks performed in python before they're eventually sent here.

mtreinish replied:
Ok, that makes sense. My comment was more about the method name and the docstring identical to the python one. This is really a different method from the python space `add_instruction` it's a subset of the functionality for handling the rust data but everything is showing it looks like the python space.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+                name
+            )));
+        }
+        let mut qargs_val: PropsMap = PropsMap::default();
```
Could we use `with_capacity()` here since  this will be unused or have `properties.len()`?

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+                                ));
+                            qarg_obj = Some(qarg.clone())
+                        }
+                        qargs_val.insert(qarg_obj.to_owned(), properties[qarg].clone());
```
This clone seems a bit weird, since we own `properties` I would expect us to be able to just move the data into `gate_map` directly.

raynelfss replied:
I will use `IndexMap::swap_remove()` to avoid cloning.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    ///     set: The set of qargs the gate instance applies to.
+    #[pyo3(name = "qargs_for_operation_name")]
+    fn py_qargs_for_operation_name(&self, operation: String) -> PyResult<Option<Vec<Qargs>>> {
+        match self.qargs_for_operation_name(&operation) {
```
I think this should work: 
```suggestion
    fn py_qargs_for_operation_name(&self, operation: &str) -> PyResult<Option<Vec<Qargs>>> {
        match self.qargs_for_operation_name(operation) {
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    ///     operations.
+    #[pyo3(name = "operation_from_name")]
+    fn py_operation_from_name(&self, instruction: String) -> PyResult<TargetOperation> {
+        match self.operation_from_name(&instruction) {
```
```suggestion
    fn py_operation_from_name(&self, instruction: &str) -> PyResult<TargetOperation> {
        match self.operation_from_name(instruction) {
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    /// Returns an iterator over the operation names in the target.
+    pub fn operation_names(&self) -> impl Iterator<Item = &String> {
+        self.gate_map.keys()
+    }
```
`&String` is a bit weird I'd expect `&str` 
```suggestion
    pub fn operation_names(&self) -> impl Iterator<Item = &str> {
        self.gate_map.keys()
    }
```

But I'm not sure the compiler is smart enough to map that, so you might need:

```suggestion
    pub fn operation_names(&self) -> impl Iterator<Item = &str> {
        self.gate_map.keys().map(|x| x.as_str())
    }
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+            let qarg_sample = qargs_keys.iter().next().cloned();
+            if let Some(qarg_sample) = qarg_sample {
+                if !strict_direction {
+                    let mut qarg_set: HashSet<SmallVec<[PhysicalQubit; 2]>, RandomState> =
```
I might rename this `deduplicated_qargs` because it wasn't entirely clear what the purpose was here. I assume it's for deduplication because otherwise `qarg_len` will always be `qargs_keys.len()`. That being said I wonder if we can move this definition outside the loop and reuse a single allocation for each instruction.

raynelfss replied:
Will rename, and I think it is already a single allocation for each instruction as these deduplicated `qargs` are those that are part of the instruction's property map.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    pub fn operation_names_for_qargs(
+        &self,
+        qargs: &Option<Qargs>,
+    ) -> Result<HashSet<&String, RandomState>, TargetKeyError> {
```
I would use `&str` here instead of `&String` also if you're using hashbrown you don't need to set `RandomState` since that's the default (which is a big part of why we use it).

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    /// Returns rust-native operation instances present in the Target that affect the provided qargs.
+    pub fn ops_from_qargs(
+        &self,
+        qargs: &Option<Qargs>,
```
It's a bit more idiomatic to use `Option<&Qargs>` because you can convert from `&Option<T>` -> `Option<&T>` with `Option.as_ref()` but you can't convert the other way.

raynelfss replied:
This will be a bit tricky to implement as it will require us to perform cloning inside of the function. I will leave this item open for further discussion.

raynelfss replied:
I implemented an alternative structure that wraps around the `IndexMap` and can process `Option<&K>` as an acceptable index.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    /// Gets all the qargs used by the specified operation name. Rust native equivalent of ``BaseTarget.qargs_for_operation_name()``
+    pub fn qargs_for_operation_name(
+        &self,
+        operation: &String,
```
I would use `&str` here

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+        &self,
+        operation_name: &String,
+        qargs: &Option<Qargs>,
+        parameters: &Option<Vec<Param>>,
```
Same comments here about argument typing from the earlier methods.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1285 @@
+    /// Retrieves an iterator over the property maps stored within the Target
+    pub fn values(&self) -> Values<String, PropsMap> {
+        self.gate_map.values()
+    }
```
Instead of using `Keys` and `Values` here is there a reason to not just use `impl Iterator` like you've done for the other APIs?

mtreinish reviewed (commented):
I did a more complete pass through the code, I think this is getting closer thanks for sticking with this.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Corrections from Matthew's review
- Use `format!` for repr method in `InstructionProperties`
- Rename `Variable` variant of `TargetInstruction` to `Variadic`.
- Remove `internal_name` attribute from `TargetOperation`.
- Remove `VariableOperation` class.
- Use `u32` for `granularity`, `pulse_alignment`, and `acquire_alignment`.
- Use `Option` to store nullable `concurrent_measurements.
- Use `&str` instead of `String` for most function arguments.
- Use `swap_remove` to deallocate items from the provided `properties` map in `add_instruction`.
- Avoid cloning instructions, use `to_object()` instead.
- Avoid using `.to_owned()`, use `.clone()` instead.
- Remove mention of `RandomState`, use `ahash::HashSet` instead.
- Move parameter check to python in `instruction_supported`.
- Avoid exposing private attributes, use the available ones instead.
- Filter out `Varidadic` Instructions as they're not supported in rust.
- Use peekable iterator to peak at the next qargs in `generate_non_global_op_names`.
- Rename `qarg_set` to `deduplicated_qargs` in `generate_non_global_op_names`.
- Return iterator instances instead of allocated `Vec`.
- Add `python_compare` and `python_is_instance` to perform object comparison with objects that satisfy the `ToPyObject` trait.
- Other small tweaks and fixes.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

raynelfss committed:
Implement a nullable dict-like structure for IndexMap (#2)

* Initial: Implement a nullable dict-like structure for IndexMap

* FIx: Erroneous item extraction from Python
- Fix error that caused `None` values to be ignored from `None` keys.
- Removed mutability from rust function argument in `add_instruction`.
   - Object is mutably referenced after option unwrapping.
- Add missing header in `nullable_index_map.rs`.
- Add Clone as a `K` and/or `V` constraint in some of the iterators.
- Remove `IntoPy` constraint from `NullableIndexMap<K, V>`.
- Add `ToPyObject` trait to `NullableIndexMap<K, V>`.

* Fix: inplace modification of Python dict.
- Perform `None` extraction from rust.
- Revert changes to `Target.py`

* Fix: Avoid double iteration by using filter_map.

* Docs: Add inline comments.

* Fix: More specific error message in `NullableIndexMap`

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'Qiskit:main' into move-target

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1298 @@
+    pub concurrent_measurements: Option<Vec<Vec<PhysicalQubit>>>,
+    gate_map: GateMap,
+    #[pyo3(get)]
+    _gate_name_map: IndexMap<String, TargetOperation, RandomState>,
+    global_operations: IndexMap<u32, HashSet<String>, RandomState>,
+    variable_class_operations: IndexSet<String, RandomState>,
+    qarg_gate_map: NullableIndexMap<Qargs, Option<HashSet<String>>>,
+    non_global_strict_basis: Option<Vec<String>>,
+    non_global_basis: Option<Vec<String>>,
```
This is more a comment for a future performance optimization to investigate in a follow up PR (we can convert this to a tracking issue after the PR merges)

All the `String` entries here are all the same set of strings. Ideally we should have only one owned string copy as the canonical name and the others are all just `&str` references to that single string. We might have issues with lifetimes on this as a pyclass, but we could do something like have a `gate_names: Vec<String>` and then replace all the strings by a usize index.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1298 @@
+    fn num_qubits(&self) -> u32 {
+        match &self {
+            Self::Normal(normal) => normal.operation.num_qubits(),
+            Self::Variadic(_) => 0,
```
I think we should either use `u32::MAX` here or `unreachable!()`. This 0 still seems off to me since from the target perspective there are no qubit constraints on a variadic operation.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1298 @@
+interface may evolve over time as the needs of the compiler change. These
+changes will be done in a backwards compatible and controlled manner when
+they are made (either through versioning, subclassing, or mixins) to add
+on to the set of information exposed by a target.
```
Can you put a description of how `BaseTarget` is defined and used here? I think it's important to document what this class is and not just repeat the Python class's docstring.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1298 @@
+            .to_object(py)
+    }
+
+    // Class properties
```
```suggestion
    // Instance attributes
```

or `properties` are fine too. But they're not class properties because they're defined for each object and not at the class level.

mtreinish commented on crates/accelerate/src/target_transpiler/nullable_index_map.rs:
```diff
@@ -0,0 +1,455 @@
+/// **Warning:** This is an experimental feature and should be used with care as it does not
+/// fully implement all the methods present in `IndexMap<K, V>` due to API limitations.
+#[derive(Debug, Clone)]
+pub struct NullableIndexMap<K, V>
```
With this warning do we want to make this fully public, or can we constrain this to `pub(crate)`? This is really only used internally for the Target right, so we may return it to python (but that doesn't matter for this) and we can avoid exposing this directly for accessing the target from rust right? (although all the usage right now is potentially coming from the same crate)

raynelfss replied:
Well, with this one it's a little complicated. Even though it is currently being used only in `Target`, it might be usable for other things in the future where we have to store a nullable value inside of a dict-like structure. And maybe someone could build on top of it.

However, the main reason why I made it public is that if we call `Target.values` the value that is returned is an instance of `NullableIndexMap` and, if we wanted to avoid that, we'd have to reconstruct the `IndexMap` entirely. Of course, there might be a better way to approach this so I'm curious as to which ideas you might have.

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -22,15 +22,19 @@
 import rustworkx as rx
 
+# import target class from the rust side
+from qiskit._accelerate.target import (  # pylint: disable=unused-import
```
Was pylint actually flagging these as unused?

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -135,8 +149,16 @@ def __repr__(self):
+
 
-class Target(Mapping):
+class Target(BaseTarget):
```
I feel like dropping `Mapping` here might be a breaking api change. Can we add this as a metaclass or something so we've registered the class as a mapping in the same way.

mtreinish reviewed (commented):
Overall I think this is about ready to merge. I think there will definitely be an interative process where we refine this as we start using it more. But this has all the pieces we need to start and only will cost a bit extra memory for a rust space copy of the properties. I left a few inline comments, but the big things I think we must fix before moving forward here are the docstrings for `BaseTarget` should describe `BaseTarget`'s functionality and what/how it is to be used. Even though it's a private class and only for Qiskit internals having the docstrings read like the python space target is confusing.

The other big one is I think dropping `Mapping` from the Python space `Target` is a breaking api change and we should be able to avoid that, I think `metaclass=Mapping` or something similar will work to fix that and let you inherit from `BaseTarget` and still register the class as a `Mapping`.

raynelfss committed:
Fix: Use `Mapping` as the metaclass for `Target`
- Minor corrections from Matthew's review.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Make `Target` crate-private.
- Due to the private nature of `NullableIndexMap`, the `Target` has to be made crate private.
- Add temporary`allow(dead_code)` flag for the unused `Target` and `NullableIndexMap` methods.
- Fix docstring of `Target` struct.
- Fix docstring of `add_instruction`.
- Make several python-only operations public so they can be used with other `PyClass` instances as long as they own the gil.
- Modify `py_instruction_supported` to accept bound objects.
- Use rust-native functions for some of the instance properties.
- Rewrite `instruction` to return parameters as slice.
- `operation_names` returns an `ExactSizeIterator`.
- All rust-native methods that return an `OperationType` object, will return a `NormalOperation` instance which includes the `OperationType` and the parameters.

raynelfss committed:
Merge branch 'main' into move-target

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1253 @@
+pub(crate) struct NormalOperation {
+    pub operation: OperationType,
+    pub params: SmallVec<[Param; 3]>,
+    op_object: PyObject,
```
Isn't this largely duplicative with the other fields? Is there something we gain by storing this too? Is it just so you don't have to construct it on the fly again if a user requests the operation object for a name?

raynelfss replied:
Yes, that is exactly the purpose of this. So that we don't have to reconstruct the `Operation` object. However, I understand that Jake's changes in #12730 might reshape the way this is done. So storing the `op_object` might not be necessary after that merges, so maybe we should wait for that PR.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1253 @@
+    /// Returns an iterator over all the instructions present in the `Target`
+    /// as pair of `&OperationType`, `&SmallVec<[Param; 3]>` and `Option<&Qargs>`.
+    // TODO: Remove once `Target` is being consumed.
+    #[allow(dead_code)]
```
Another note for a follow-up we should add rust space tests for these, since they're meant to be consumed from rust we should be able to test them without python.

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -994,25 +740,32 @@ def _build_coupling_graph(self):
+                    if self.operation_from_name(gate).num_qubits == 2:
+                        self._coupling_graph = (
+                            None  # pylint: disable=attribute-defined-outside-init
+                        )
```
Is this pylint disable still needed? https://github.com/Qiskit/qiskit/pull/12292/files#diff-00276708f86c726ae04a2c670e74177eaad6aad971e017951bb5792501d13bf5R338 seems to define it in `__init__`

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -1188,10 +893,12 @@ def __str__(self):
                 if duration is not None:
-                    prop_str_pieces.append(f"\t\t\tDuration: {duration} sec.\n")
+                    prop_str_pieces.append(
+                        f"\t\t\tDuration: {duration if duration > 0 else 0} sec.\n"
```
When is `duration < 0`?

raynelfss replied:
This is so that Python doesn't display the floating point number at 0. Of course there might be a better way of doing this.

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -1188,10 +893,12 @@ def __str__(self):
                 error = getattr(props, "error", None)
                 if error is not None:
-                    prop_str_pieces.append(f"\t\t\tError Rate: {error}\n")
+                    prop_str_pieces.append(f"\t\t\tError Rate: {error if error > 0 else 0}\n")
```
Same question here when is `error < 0`

raynelfss replied:
> This is so that Python doesn't display the floating point number at 0. Of course there might be a better way of doing this.

Same here.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -0,0 +1,1253 @@
+The intent of this struct is to contain data that can be representable and
+accessible through both Rust and Python, so it can be used for rust-based
+transpiler processes.
+ */
```
The only thing I think we should mention here is around where we have copies of data or not and how we keep things in sync. Just so people will understand the split in responsibilities between the python space and the rust space. Otherwise this is much better.

mtreinish reviewed (commented):
I think this is ready to go now. I have a few very small nits inline. But for a 2k line change I think I can continue to look at this and find something forever. so after these are fixed I'll approve and we can merge this. I think there are some potential cleanups we'll want to do down the road, but this is a great start and gets us the data model for the target in rust to start leveraging in various places.

This will conflict with @jakelishman's PR #12730 I think because that removes `OperationType`  but it shouldn't be too hard to update this with that or vice versa if this merges first.

raynelfss committed:
Fix: Comments from Matthew's review
- Mention duplication in docstring for rust Target.
- Use f"{*:g}" to avoid printing the floating point for 0 in `Target`'s repr method.
- Add note mentioning future unit-tests in rust.

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Merge branch 'main' into move-target

raynelfss committed:
Fix: Adapt to #12730

mtreinish reviewed (approved):
LGTM, thanks for all the hard work on this @raynelfss. I think this is ready to merge. There are some small things I outlined in my earlier reviews that we can fix in follow ups. This is a new rust space representation so just like with everything we're migrating over there will likely be a longish tail of follow-ups we'll make to make incremental improvements as we start to use it more. But this is definitely in a state where I think it's good to start working with.

This all being said I'm going to hold off on pushing the automerge button right now, because I think we should wait until we branch 1.2.0 for this. Since nothing in 1.2.0 will use this I think it's better to just include this in 1.3.0.

IvanaGyro commented:
@raynelfss Hi, is making `BaseTarget.num_qubits` immutable intended? [Many tests in Qiskit Dynamics](https://github.com/qiskit-community/qiskit-dynamics/actions/runs/12726242543/job/36024767691?pr=368) failed because they tried to set the value of `num_qubits`. I am not sure if I have to fix all tests to let the tiny fix: https://github.com/qiskit-community/qiskit-dynamics/pull/368 be merged.

code of `BaseTarget.num_qubits`: https://github.com/Qiskit/qiskit/blame/e9ccd3f374fd5424214361d47febacfa5919e1e3/crates/accelerate/src/target_transpiler/mod.rs#L151-L152
